packages/wp-source/src/actions.ts

```diff
@@ -24,49 +24,41 @@ const actions: WpSource["actions"]["source"] = {
     const linkParams = parse(route);
     const { query, page, queryString } = linkParams;
 
-    // Get current data object.
-    const data = source.data[link];
-
     // Get options.
     const force = options ? options.force : false;
 
-    if (!data) {
-      // If there is no data yet, just set the necessary flags.
-      source.data[link] = {
+    // Get current data object.
+    let data = source.data[link];
+
+    // Initialize `data` if it does not exist yet. Also, reinitialize it only in
+    // the case `data` is an error and `{ force: true }` is used.
+    if (!data || (force && isError(data))) {
+      data = source.data[link] = {
         isFetching: true,
         isReady: false,
+        route: linkParams.route,
+        link,
+        query,
+        page,
       };
-    } else if (force) {
-      // If we fetch with `{ force: true }`, then only set `isFetching` to true
-      // again.
-      data.isFetching = true;
+    } else {
+      // Reassign `route`, `link`, `query`, `page` to fix custom handlers that
+      // do not add them.
+      data = source.data[link] = {
```

Comment:
I see. Then I should change another similar assignment in line 101, right?

Reply:
No, if the object already exists in the state, update its individual properties:

```js
source.data[link].link = link;
source.data[link].route = linkParams.route;
source.data[link].query = query;
source.data[link].page = page;
```

Or use `Object.assing`:

```js
Object.assign(source.data[link], {
  link: link,
  route: linkParams.route,
  query: query,
  page: page,
});
```

If it's not clear why we should do this let me know and I will explain it further 🙂

Reply:
No, no, I understood.

The thing is, in that line, the data object was already created and it's being recreated again. That's why I said to change it as well.

https://github.com/frontity/frontity/blob/a7f9743546/packages/wp-source/src/actions.ts#L100-L106

Reply:
OK, maybe that assignment is in purpose. :smile:

Reply:
If I'm not mistaken in that line we are moving from

```js
{
  isReady: false,
  isFetching: true
}
```

to

```js
{
  isReady: true,
  isFetching: false
}
```

so it forces a rerender anyway.

But you are right. We should try to avoid creating new objects as much as possible, even in this case because there may be a React component listening to the `data.page` for example, and that one will get rerendered unnecessarily.

Reply:
Well, I'm going to leave those lines as it is, we can change them later if needed.